File: Difficulty: Hard/LRU Cache/lru-cache.cpp

class Node {
    public:
        int key, val;
        Node* prev;
        Node* next;
        
        Node(int key,int val) {
            Node::key = key;
            Node::val = val;
            prev = nullptr;
            next = nullptr;
        }
    
};


class LRUCache {
  private:
    unordered_map<int,Node*> ump;
    int cap;
    Node* head, *tail;
    int size;
  public:
    LRUCache(int cap) {
        // code here
        LRUCache::cap = cap;
        head = new Node(0,0);
        tail = new Node(0,0);
        head->next = tail;
        head->prev = nullptr;
        tail->prev = head;
        tail->next = nullptr;
        size = 0;
    }
    
    void swap(Node* cur) {
      
     
        (cur->prev)->next = cur->next;
        (cur->next)->prev = cur->prev;
            
        cur->next = head->next;
        cur->prev = head;
        (head->next)->prev = cur;
        head->next = cur;
        
        
        
    }

    int get(int key) {
        
        int ans = -1;

        if(ump.find(key)!=ump.end()) {
            ans =  (ump[key]->val);
            /// TODO: WILL FURTHER PROCESS
            Node* cur = ump[key];
           
            swap(cur);
            
            
            
        }
        
        return ans;
        
        // code here
    }
    
    void addNode(Node* newNode) {
        newNode->prev = head;
        newNode->next = head->next;
        (head->next)->prev = newNode;
        head->next = newNode;
        ump[newNode->key] = newNode;
    }
    
    void update(int key, int value ) {
        Node* cur = tail->prev;
        
        ump.erase(cur->key);
        
        
        cur->key = key;
        cur->val = value;
        
        ump[key] = cur;
        
        swap(cur);
        
    }
        
    void put(int key, int value) {
        // code here
        if(ump.find(key)!=ump.end()) { // cache hit
            ump[key]->val = value;
            swap(ump[key]);
        }
        else {
            if(size<cap) { // if there is cache miss and still we have capacity
                size++;
                Node* newNode = new Node(key,value);
                addNode(newNode);
            }
            else {
                update(key,value); // we are reusing tail to update
            }
        }
        
        
    }
};

File: Difficulty: Hard/LRU Cache/lru-cache.java

class Node {
    public int key;
    public int value;
    
    public Node next;
    public Node prev;
    
    public Node() {}
    
    public Node(int key, int value) {
        this.key = key;
        this.value = value;
    }
}

class LRUCache {
    
    private static Map<Integer,Node> mp;
    
    private static Node head;
    private static Node tail;
    private static int cap;
    private static int size;

    LRUCache(int capi) {
        // code here
        
        mp = new HashMap<Integer,Node>();
        head = new Node();
        tail = new Node();
        
        head.next = tail;
        tail.prev = head;
        
        cap = capi;
        size = 0;
        
    }
    
    public static void swap(Node node) {
        
        if(node.prev != null && node.next != null) {
            (node.prev).next = node.next;
            (node.next).prev = node.prev;
        }
        
        
        node.next = head.next;
        node.prev = head;
        (head.next).prev = node;
        
        head.next = node;
        // if(tail.prev == head)
        //     tail.prev = node;
        
    }
    
    public static void remove() {
        Node delNode = tail.prev;
        mp.remove(delNode.key);
        Node prevNode = delNode.prev;
        prevNode.next = tail;
        tail.prev = prevNode;
        delNode.next = null;
        delNode.prev = null;
        
    }
    
    public static void add(int key, int value) {
       
        
        if(size<cap) {
            size++;
            
        } else {
            remove();
            
        }
        
        Node newNode = new Node(key,value);
        
        update(newNode,-1);
        
        mp.put(key,newNode);
        
    }
    
    public static void update(Node node, int value) {
        if(value!=-1) {
            node.value = value;
        }
        
        swap(node);
    }
    
    

    public static int get(int key) {
        //  code here
        Node res = mp.getOrDefault(key,null);
        if(res==null) return -1;
        update(res,-1);
        return res.value;
    }

        
    public static void put(int key, int value) {
        //  code here
        Node res = mp.getOrDefault(key,null);
        if(res==null) {
            add(key,value);
        }
        else {
            update(res,value);
        }
    }
}